### PR TITLE
[Hotfix] Adapt LR scheduler to epoch wise

### PR DESCRIPTION
**Pull request by nabenabe0928**
Address the [issue#211](https://github.com/automl/Auto-PyTorch/issues/211).

### CODE REVIEW

File: autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py [hunk 1]
```diff
@@ -23,39 +28,42 @@ def __init__(self, weighted_loss: bool = False,
...
         return lambda criterion, pred: criterion(pred, y_a)
 
     @staticmethod
-    def get_properties(dataset_properties: typing.Optional[typing.Dict[str, typing.Any]] = None
-                       ) -> typing.Dict[str, str]:
+    def get_properties(dataset_properties: Optional[Dict[str, Any]] = None
+                       ) -> Dict[str, str]:
         return {
             'shortname': 'StandardTrainer',
             'name': 'StandardTrainer',
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py:61]:
```suggestion
            'name': 'Standard Trainer',
```

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 2]
```diff
@@ -185,22 +203,34 @@ def prepare(
         metrics_during_training: bool,
         scheduler: _LRScheduler,
         task_type: int,
-        labels: Union[np.ndarray, torch.Tensor, pd.DataFrame]
+        labels: Union[np.ndarray, torch.Tensor, pd.DataFrame],
+        step_unit: Union[str, StepIntervalUnit] = StepIntervalUnit.batch
     ) -> None:
 
+        step_unit_types = [str, StepIntervalUnit]
+        if not any(isinstance(step_unit, step_unit_type) for step_unit_type in step_unit_types):
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:211]:
```suggestion
        if isinstance(step_unit, step_unit_types):
```
you can check if an object is in an iterable of types directly.

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 3]
```diff
@@ -185,22 +203,34 @@ def prepare(
...
+            raise ValueError(f'step_unit must be {step_unit_types}, but got {type(step_unit)}.')
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:212]:
```suggestion
            raise ValueError(f'step_unit must be one of {step_unit_types}, but got {type(step_unit)}.')
```

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 4]
```diff
@@ -185,22 +203,34 @@ def prepare(
...
+
         # Save the device to be used
         self.device = device
 
         # Setup the metrics
         self.metrics = metrics
 
         # Weights for the loss function
-        kwargs = {}
+        weight_dict = {}
```
Comment (ravinkohli) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:228]: can we use weight_params? I think that would explain it better.

**Review by ravinkohli — CHANGES_REQUESTED** [review]
Thank you for this PR. I have made a few suggestions. Also, do you think in the scheduler tests, we can check if the step of the scheduler is taken at the expected time. However, I could not find where the scheduler is passing its step interval anywhere? What am I missing?

**Comment by nabenabe0928** [comment]
> I have made a few suggestions. 

Thanks, I will reflect those changes:)

> Also, do you think in the scheduler tests, we can check if the step of the scheduler is taken at the expected time. 

Do you mean we want another test for it? (Then I will do it.)

> However, I could not find where the scheduler is passing its step interval anywhere? What am I missing?

Since everything is set in `set_training_params`, I put it there.

File: autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py (hunk 1 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py:61]: Why do we want space here?

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py (hunk 2 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/trainer/base_trainer.py:211]: Ok, I will change to the following:
```
if not isinstance(step_unit, step_unit_types):
```

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py (hunk 3 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/trainer/base_trainer.py:212]: I will change it to `either`.

File: autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py (hunk 1 above)
Comment (ravinkohli) [reply autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py:61]: its not important, but in general we have followed that short name is either acronym of the whole name or the whole name without spaces. You can definitely ignore this but I suggested for uniformity

**Comment by ravinkohli** [comment]
> > I have made a few suggestions.
> 
> Thanks, I will reflect those changes:)
> 
> > Also, do you think in the scheduler tests, we can check if the step of the scheduler is taken at the expected time.
> 
> Do you mean we want another test for it? (Then I will do it.)
> 
> > However, I could not find where the scheduler is passing its step interval anywhere? What am I missing?
> 
> Since everything is set in `set_training_params`, I put it there.
but how is the scheduler telling the base trainer that it should be updated epoch or batch wise? I see that by default you are passing `batch` but its not connected to the scheduler right? For this to work, we should send this info via base_trainer_choice, which will pass it via `step_unit` of the `set_training_params`.

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py [hunk 5]
```diff
@@ -236,11 +273,25 @@ def on_epoch_end(self, X: Dict[str, Any], epoch: int) -> bool:
...
+    def _scheduler_step(
+        self,
+        step_interval: StepIntervalUnit,
+        loss: Optional[torch.Tensor] = None
+    ) -> None:
+        if self.step_unit != step_interval:
+            return
+
+        if self.scheduler:
+            if 'ReduceLROnPlateau' in self.scheduler.__class__.__name__:
```
Comment (franchuterivera) [inline autoPyTorch/pipeline/components/training/trainer/base_trainer.py:285]: I think it will be better not to hardcode things in this, and let every scheduler decide what type of step it should take?

File: autoPyTorch/pipeline/components/training/trainer/MixUpTrainer.py [hunk 6]
```diff
@@ -29,45 +32,50 @@ def __init__(self, alpha: float, weighted_loss: bool = False,
...
 
         mixed_x = lam * X + (1 - lam) * X[index, :]
         y_a, y_b = y, y[index]
-        return mixed_x, {'y_a': y_a, 'y_b': y_b, 'lam': lam}
-
-    def criterion_preparation(self, y_a: np.ndarray, y_b: np.ndarray = None, lam: float = 1.0
-                              ) -> typing.Callable:
+        return mixed_x, _NewLossParameters(y_a=y_a, y_b=y_b, lam=lam)
+
+    def _get_new_loss_fn(
```
Comment (franchuterivera) [inline autoPyTorch/pipeline/components/training/trainer/MixUpTrainer.py:59]: Can we try to make this in another PR? I think this does not directly helps to solve the `LR schedule`

**Comment by franchuterivera** [comment]
Thanks a lot for the changes, for example, there are places where you coded certain pieces in a better way! Can these changes be done on a separate PR? I think this will be checked and approved right away as they are small format changes.

However, they make the actual change quite complicated to surface -- I also do not see right away how every scheduler will define whether the step should be done per epoch or per batch? 

Do you think we can also add unit testing to prevent this from happening again?

File: autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py (hunk 1 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/trainer/StandardTrainer.py:61]: Ok, I do not know where we are using those names, but if it is convention, I will follow this rule.

File: autoPyTorch/pipeline/components/training/trainer/base_trainer.py (hunk 5 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/trainer/base_trainer.py:285]: Yeah, I know what you mean, but it was the original code and I cannot guarantee the original behavior by changing here, so I would like to keep this for now.
Of course, if you have any good idea to solve this, I may adapt here immediately.

File: autoPyTorch/pipeline/components/training/trainer/MixUpTrainer.py (hunk 6 above)
Comment (nabenabe0928) [reply autoPyTorch/pipeline/components/training/trainer/MixUpTrainer.py:59]: The renaming? Ok, I will do it in another PR.

**Comment by ravinkohli** [comment]
As we have solved this issue with #256, I am closing this PR
